### PR TITLE
Add: 376 update events

### PR DESCRIPTION
**Pull request by L03TJ3**
# Description
To catch the onPress event of the web3action button, there must be a way to send events back up.
This adds an eventsCb prop so that the different stages of the web3action flow could be watched/or catched
to notify using app 

About # (link your issue here)
https://github.com/GoodDollar/GoodProtocolUI/issues/376

### CODE REVIEW

File: packages/good-design/src/advanced/web3action/Web3Action.tsx [hunk 1]
```diff
@@ -5,15 +5,20 @@ import BaseButton, { BaseButtonProps } from "../../core/buttons/BaseButton";
...
   switchChain?: (requiredChain: number) => Promise<any>;
   handleConnect?: () => Promise<any> | void;
+  /** callback to be used for receiving the events: start,finish,connect,switch,action
+   *
+   * 'connect/switch' will send suffixed 'start' and 'success'
+   *
+   * 'action' only has a suffixed '_start',
+   * final 'finish' indicating action has been done
+   */
+  eventsCb?: (event: string) => void;
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/advanced/web3action/Web3Action.tsx:21]:
```suggestion
  onEvent?: (event: string) => void;
```

File: packages/good-design/src/advanced/web3action/Web3Action.tsx [hunk 2]
```diff
@@ -49,81 +54,97 @@ const StepIndicator: FC<{ text?: string } & ITextProps> = withTheme({ name: "Ste
...
-
-  const switchToChain = useCallback(
-    async (chain: number) => {
-      const switchFn = switchChain || switchNetwork;
-      const result = await switchFn(chain).catch(throwIfCancelled);
-
-      if (switchChain && !result) {
-        throwCancelled();
+})(
+  ({ text, supportedChains, switchChain, web3Action, handleConnect, eventsCb, innerIndicatorText, ...buttonProps }) => {
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/advanced/web3action/Web3Action.tsx:58]:
```suggestion
  ({ text, supportedChains, switchChain, web3Action, handleConnect, onEvent = noop, innerIndicatorText, ...buttonProps }) => {
```

File: packages/good-design/src/advanced/web3action/Web3Action.tsx [hunk 3]
```diff
@@ -49,81 +54,97 @@ const StepIndicator: FC<{ text?: string } & ITextProps> = withTheme({ name: "Ste
...
+
+    const resetText = useCallback(() => setActionText(""), []);
+
+    const sendEvents = useCallback(
+      (event: string) => {
+        if (!eventsCb) return;
+        eventsCb(event);
+      },
+      [eventsCb]
+    );
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/advanced/web3action/Web3Action.tsx:72]: no need 
```suggestion
```

File: packages/good-design/src/advanced/web3action/Web3Action.tsx [hunk 4]
```diff
@@ -49,81 +54,97 @@ const StepIndicator: FC<{ text?: string } & ITextProps> = withTheme({ name: "Ste
...
+
+    const finishFlow = useCallback(() => {
+      resetText();
+      setRunningFlow(false);
+      sendEvents("finish");
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/advanced/web3action/Web3Action.tsx:77]:
```suggestion
      onEvent("finish");
```

**Review by johnsmith-gooddollar — CHANGES_REQUESTED** [review]
optionally define events names as constants, you could use TS enum for that (don't forget to adjust types)

File: packages/good-design/src/advanced/web3action/Web3Action.tsx (hunk 1 above)
Comment (L03TJ3) [reply packages/good-design/src/advanced/web3action/Web3Action.tsx]: done

File: packages/good-design/src/advanced/web3action/Web3Action.tsx (hunk 2 above)
Comment (L03TJ3) [reply packages/good-design/src/advanced/web3action/Web3Action.tsx]: done

File: packages/good-design/src/advanced/web3action/Web3Action.tsx (hunk 3 above)
Comment (L03TJ3) [reply packages/good-design/src/advanced/web3action/Web3Action.tsx]: Ah yes, pretty rendunant :)

File: packages/good-design/src/advanced/web3action/Web3Action.tsx (hunk 4 above)
Comment (L03TJ3) [reply packages/good-design/src/advanced/web3action/Web3Action.tsx]: done

**Comment by L03TJ3** [comment]
@johnsmith-gooddollar the optional comment you made
do we want to implement this immediately? 
Or do we want to take a little more time (separate issue) to implement a more generic approach across good-design for events?

**Pull request merged**